### PR TITLE
[region-isolation] Fix actor region propagation and some other smaller issues.

### PR DESCRIPTION
**Pull request by gottesmm**
This PR contains a few different changes:

1. It teaches the pass how to handle actor regions. Put simply, we in a flow insensitive way identify if specific values are isolated to an actor and then if a region contains any value with an actor isolated value, then the whole region must be actor isolated. Thus we can take advantage of the region dataflow to make our flow insensitive constraint flow sensitive.
2. A hack was added some time ago that made it so that if an address was passed to any address, then we needed to make it so that assignments were actually merges (meaning that later assignments would not erase the old region a value belonged to). This should only semantically happen with partial_apply. As far as the codegen of the compiler is concerned, I was unable to come up with a case where the unnecessary merges were harmful. I could construct something with SIL but unfortunately we can't write SIL test cases yet with this pass.... = /.
3. I noticed that the whole cache vector of argument IDs for function arguments was only used to construct a Partition for the entry block. Rather than having this extra concept, I changed the function arguments to just initialize directly the partition... eliminating code.
4. I noticed that we were not properly handling begin_access. Specifically, begin_access was being treated as an assignment. That means that if one were to transfer a var and then attempt to reinitialize the var, the begin_access of the reinitialization would be considered a use of the transferred value and we would emit an error... which is incorrect. Instead, we just look through begin_access.
5. The pass as originally written is a pass that does a bunch of work deep in the callstack that doesn't need to be done so deep. By making it so deep, it obscures what the pass is actually trying to do. An example of this is that before this PR, we initialized a block's contained pseudo-IR lazily when we were performing dataflow. This is not necessary and makes it non-obvious where the block's translation is happening. Instead we just perform it for all blocks when we construct their state and before we perform the dataflow. As a bonus, this makes the logging easier to understand since we first translate all blocks to the pseudo-ir and then perform the dataflow instead of interleaving the two.

That being said, I found a few bugs in the process that I filed radars for that I will fix in subsequent PRs. I added tests for all of them into this PR with TODOs.

1. I discovered that in certain cases, I can get SILGen to initialize aggregate values separately in pieces. If a later value is a Sendable type that is fresh, we lose the original region of the value. This can cause us to swallow diagnostics. The solution is to make it so that if we are assigning into only one part of an aggregate, we merge into the region so we don't forget the original region. rdar://117273952
2. I discovered that when one returns a non-Sendable value, if we were going to emit an error on the return (since the return would be a use after a transfer lets say), then we do not have an expression associated with the value so we crash. This is why we really should be using SILLocations instead so we get proper diagnostics. rdar://117438350
3. I discovered that we are not properly performing a union operation when combining regions from different blocks. rdar://117437059.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SILOptimizer/Mandatory/SendNonSendable.cpp [hunk 1]
```diff
@@ -857,6 +914,16 @@ class PartitionOpTranslator {
     case SILInstructionKind::SuperMethodInst:
     case SILInstructionKind::ObjCSuperMethodInst:
       return translateSILAssign(inst->getResult(0), inst->getOperand(0));
+    case SILInstructionKind::BeginBorrowInst: {
+      return translateSILAssign(inst->getResult(0), inst->getOperand(0));
+    }
+
+    case SILInstructionKind::EnumInst: {
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/SendNonSendable.cpp:944]: Is this case new? Does it have a test?

File: lib/SILOptimizer/Mandatory/SendNonSendable.cpp [hunk 2]
```diff
@@ -339,7 +339,7 @@ class PartitionOpTranslator {
   std::vector<TrackableValueID> neverTransferredValueIDs;
 
   /// A cache of argument IDs.
-  SmallVector<TrackableValueID> argIDs;
+  std::optional<Partition> functionArgPartition;
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/SendNonSendable.cpp:342]: Maybe make it mutable?

**Review by slavapestov — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/SendNonSendable.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/SendNonSendable.cpp:944]: Yes. It came up with the enum test cases I added below. I wish we had SIL test cases which would make it easier to understand the connection.

**Pull request merged**